src/Chainweb/Pact/TransactionExec.hs

```diff
@@ -627,14 +632,48 @@ findPayer = runMaybeT $ do
     mkApp i r as = App (TVar r i) (map (liftTerm . fromPactValue) as) i
 
     runCap i capRef as input = do
-      ar <- evalCap i CapCallStack False $ mkApp i capRef as
+      msgBody <- enrichedMsgBody cmd
+      ar <- local (setEnvMsgBody msgBody) $
+        evalCap i CapCallStack False $ mkApp i capRef as
+
       case ar of
         NewlyAcquired -> do
           r <- input
           popCapStack (const (return ()))
           return r
         _ -> evalError' i "Internal error, GAS_PAYER already acquired"
 
+enrichedMsgBody :: Command (Payload PublicMeta ParsedCode) -> Eval e Value
```

Comment:
This looks like it can be pure no?

Reply:
Yep, the `Eval` was a left over from when I was attempting to read from the state. Fixed in the upcoming commits.